src/Lecture2.hs

```diff
@@ -268,7 +330,23 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval progScopeVars expr = case expr of 
+  Lit val -> Right val
+  Var varName -> 
+                let 
+                  varVal = lookup varName progScopeVars
+                in case varVal of
+                    Just n -> Right n
+                    Nothing -> Left (VariableNotFound varName)
+  Add left right -> 
+                  let
+                    leftEval  = eval progScopeVars left
+                    rightEval = eval progScopeVars right
+                  in case leftEval of
+                    Left (VariableNotFound _) -> leftEval
+                    Right x -> case rightEval of
+                          Left (VariableNotFound _) -> rightEval
+                          Right y -> Right (x + y)
```

Comment:
That's a great pattern matching here! 👏🏻 
Once you learn about Monads, you'll be able to reduce this function to a single line ✂️